Departments

ianhayes opened the pull request:
Hey @buob when you get a chance can you look into this for us: We are having issues saving the users department on signup. We are getting this in terminal when we use binding.pry

`"user"=>{"first_name"=>"fdgs", "last_name"=>"dfgsdfsg", "email"=>"dsfgfdsg@gdfdsg.com", "password"=>"[FILTERED]", "password_confirmation"=>"[FILTERED]"}, "department_id"=>{"department_id"=>"1"}, "commit"=>"Sign up"}`

Is this because it is trying to input a hash?

buob commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -14,10 +14,9 @@
 
   <div class="field">
-    <%= f.fields_for :department do |d| %>
+    <%= f.fields_for :department_id do |d| %>
```
Yeah, it's definitely because of the hash being submitted, and the hash is happening because you're doing the `fields_for` function.  If you don't do that, and just do `f.label :department_id`, it should work.

Let me know how that works for you.

ianhayes commented:
Hey @buob I dropped the `fields_for` function and changed `d.label :department` to `f.label :department_id` and I am still getting `"department_id"=>{"department_id"=>"1"}`.

buob commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -14,11 +14,8 @@
-          Department::DEPARTMENT, class: "department dropdown"%>
-    <% end %>
+      <%= f.label :department_id %>
+      <%= collection_select(:department_id, :department_id, Department.all, :id, :name) %>
```
do you know why there are two `:department_id`s here, and what happens when you take one out?  Looking at the docs, the example only has one key

ianhayes replied:
Yeah, the options we need to use are `object, method, collection, value_method, text_method` based off those options on form submission we are returned with `"object"=>{"method"=>":id"}`... We used `collection_select` because we needed to pull in the options for the select from the departments table in the database. Maybe we should use a different method here instead that could do the same thing that wouldn't submit a hash? Or is there a way to add only the value from the hash to the column in that table of the db? Also, we tried removing and changing the options and its mandatory to have those 5.

buob replied:
you can definitely dissect the hash on the server, but it doesn't make much sense to me to have an object, because it's not a separate entity, right?  there's a field on the user table for department_id?  Or at least that's how it should be, otherwise you'll have to create a new department for each user.

ianhayes replied:
That's right, there is a field on the user table for department_id. I know we just want to send the integer of the department_id. I don't think using collection_select is what we're looking for in this case, based off what you're saying? What should we use instead to display the select items and only send the id rather than the full hash or send an object?

buob replied:
I haven't worked with rails helpers in a while, but looking up form helpers page and scrolling to select, in the [option tags from a collection of arbitrary objects section](http://guides.rubyonrails.org/form_helpers.html#option-tags-from-a-collection-of-arbitrary-objects), it looks like you'll want something like...

```ruby
<%= f.collection_select(:department_id, Department.all, :id, :name) %>
```

Does that work?

ianhayes replied:
That did the trick. Thanks!

ianhayes commented:
Hey @buob, Thanks for the help previously. Now that we have that fixed, when we go to our profile show page we want to display the name of the department the user belongs to. We've managed to figure out how to display the id using `@user.department_id`.. but I tried doing @user.departments.name but we are getting `No Profile Method error in #show`. I'm not entirely sure how we should add this to the show method in the controller. Any help would be great.

buob commented:
hey hey, so any method/function that is available on a `@user` object comes from that object's model declaration.  So looking at the user model declaration, a user `has_one :department` (not multiple), so the method should be `@user.department`, and then calling `.name` on that should work.

Try that out and let me know if that works.

ianhayes commented:
Hey, I just tried doing `@user.department.name` and getting the error `ActiveRecord::StatementInvalid in MyProfile#show`. I'm not quite sure how to go about fixing this.

buob commented:
You'll probably want to take a look at the [difference between `belongs_to` and `has_one`](http://requiremind.com/differences-between-has-one-and-belongs-to-in-ruby-on-rails/) on that user/department relationship.  That relationship is going to construct the SQL to look up the department based on the user, so if the relationship is wrong, the SQL is going to be wrong, which ActiveRecord is running.

ianhayes commented:
@buob thanks for sending that link over, that really clears things up!
